add ch_listen to listen socket

mattn opened the pull request:
This change add new function ch_listen() to listen socket channel. To know more description, please read my slide(P79-P81) of keynote speach on VimConf2018.

mattn committed:
support ch_listen

mattn committed:
update doc

mattn committed:
support blob

mattn committed:
support blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
listen

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
Merge branch 'master' into blob

mattn committed:
Merge branch 'master' into listen

mattn committed:
Merge branch 'master' into listen

mattn committed:
Merge branch 'master' into blob

mattn committed:
fix index()

mattn committed:
fix blob

mattn committed:
fix blob

mattn committed:
fix blob

mattn committed:
let blob[0:1]

mattn committed:
Merge branch 'master' into blob

mattn committed:
Merge branch 'master' into listen

mattn committed:
Merge branch 'blob' of bitbucket.org:mattn/vim into blob

mattn commented:
IMO, This is useful to communicate to Vim(s) with some protocols. For example, we will be possible to implement something without +clientserver feature.

lilydjwg commented:
Would you add UNIX domain socket support? With TCP sockets you'd need a way to authenicate clients or trust all users on the machine (including all the daemon users for services). This is why I don't use plugins like deoplete.

mattn commented:
Okay, I'll add it that specifying file path to the socket file with `unix:` prefix.

mattn committed:
Merge branch 'blob' of github.com:mattn/vim into blob

mattn committed:
fix blob

mattn committed:
add test

mattn committed:
fix Makefile

mattn committed:
fix Make_mvc.mak

mattn committed:
Fix doc

mattn committed:
Fix compilation warnings

mattn committed:
Support AF_UNIX

This support Windows 10 too

mattn committed:
Merge branch 'listen' of github.com:mattn/vim into listen

mattn commented:
Now supported AF_UNIX. This should work on Windows 10 too.

mattn committed:
fix build

mattn committed:
fix build

mattn committed:
fix build

mattn committed:
fix build

h-east committed:
I made the following modifications.
- The compiler warning has been removed.
- Fixed blob.c being specified twice in BASIC_SRC in Makefile.
- We added the file addition position in the Makefile as much as possible to the alphabet order.
- Changed comment style of added code to "//".

mattn committed:
Merge pull request #1 from h-east/blob

I made the following modifications.

mattn committed:
support blob in if_python3

mattn committed:
add String#to_blob to return string 0zXXXX in if_ruby

mattn committed:
Merge branch 'listen' of github.com:mattn/vim into listen

mattn committed:
Merge branch 'master' into listen

mattn committed:
Merge master

mattn committed:
Merge master

mattn committed:
Merge branch 'master' into listen

mattn committed:
Fix build

mattn committed:
Merge branch 'master' into listen

brammool commented:
This requires more documentation.  A practical example is preferred.

brammool commented:
Still awaiting  a practical example.

prabirshrestha commented:
Here is an example I can think of:

Vim is single threaded so even if we improve perf by vimscript9 or lua there will be certain heavy computation that will cause vim to pause for certain time. Being able to start a new vim process and communicate with sockets/pipes/stdin/stdout will help us. My previous ask on having remote lua plugins contains more details on a concrete example such as fuzzy matching. https://github.com/vim/vim/issues/2115.
https://github.com/natebosch/vim-lsc/issues/291 Here was another example where dart language server was returning 27k items and vim was hanging while vscode had no issue.

yegappan commented:
Another example use case for ch_listen() function:
Implementing a debugger for Vim scripts. We can run the script under debug in a
separate Vim instance where a server is running (using ch_listen). From the outside
Vim instance, we can connect to this server and run various commands (using the
JSON interface) to get the state of the script. Currently we need to use an external
server (written in Python or some other language) for this. Using ch_listen(), we can
natively do this in Vim without any external dependencies.

puremourning commented:
@yegappan did you see https://groups.google.com/forum/?nomobile=true#!searchin/vim_dev/Vimspector$20/vim_dev/Xzb0a3bTg1U/m3WMcOu2BAAJ  done without listen  uses a debug adapter.

brammool commented:
I can guess ch_listen() can make some things  work that were not possible before.  But this pull request is still missing an example.  When someone wants to use ch_listen(), it must be clear how it works.  Also to understand that it cannot do.  A working example that someone can play with works best.

vim-ml commented:
Hi Ben,

On Sun, Jun 14, 2020 at 3:30 AM Ben Jackson <vim-dev-github@256bit.org>
wrote:

> @yegappan <https://github.com/yegappan> did you see this?
> https://groups.google.com/forum/m/#!searchin/vim_dev/Vimspector$20/vim_dev/Xzb0a3bTg1U
> done without listen uses a debug adapter.
>
>
> I went through the Vimspector documentation. It looks like it supports
languages
with a debug adapter but doesn't support debugging Vimscript. I was
referring to
creating a debugger for Vimscript. Even with ch_listen(), we will need
support
for running the Vim debugger in the background while still being able to see
the Vim output screen. This will need some changes to the current Vim
debugger.

Regards,
Yegappan

puremourning commented:
@yegappan 

>  wasreferring to creating a debugger for Vimscript. 

that's exactly what I've done, and what my RFC and demo are all about: https://github.com/puremourning/vim-debug-adapter and my vim debugger branch: https://github.com/puremourning/vim/tree/debugger


I think I might have copied a dud url - this is the RFC I sent: https://groups.google.com/d/msg/vim_dev/Xzb0a3bTg1U/m3WMcOu2BAAJ

vim-ml commented:
Hi,

On Sun, Jun 14, 2020 at 12:33 PM Ben Jackson <vim-dev-github@256bit.org>
wrote:

> @yegappan <https://github.com/yegappan>
>
> wasreferring to creating a debugger for Vimscript.
>
> that's exactly what I've done, and what my RFC and demo are all about:
> https://github.com/puremourning/vim-debug-adapter and my vim debugger
> branch: https://github.com/puremourning/vim/tree/debugger
>
> I think I might have copied a dud url - this is the RFC I sent:
> https://groups.google.com/d/msg/vim_dev/Xzb0a3bTg1U/m3WMcOu2BAAJ
>
>
>
This looks very interesting and useful. Hopefully this gets integrated into
Vim at some point.

Regards,
Yegappan

puremourning commented:
Thanks. My goal is to get it all to a shippable/professional level "in 2020". I expect the first PRs in Q3. It's likely to be a lot of smaller PRs and some discussion with Bram/etc. so I wouldn't expect it super soon, but you're more than welcome to try the branch out now. It does _mostly_ work.

lambdalisue commented:
Any news? I'd love to make a plugin which shows a prompt/buffer to input text for commands in `job_start()` through some environment variable (e.g. `EDITOR` or `SSH_ASKPASS`).

I can make it for Neovim while it has RPC but Vim. If there is `ch_listen()`, I can use it to handle the Vim from a process executed in `job_start()`.

mattn committed:
Merge branch 'master' into listen

brammool commented:
This is still lacking an example.
The help is very brief.  I guess "List to a channel" actually means "listen on a socket".
Does this work on all systems?
"sys/un.h" is a newly used include file, is this available everywhere or does it require a configure check?
The test lacks testing for the possible errors.

puremourning commented on runtime/doc/channel.txt:
```diff
@@ -584,6 +584,17 @@ ch_info({handle})						*ch_info()*
 		Can also be used as a |method|: >
 			GetChannel()->ch_info()
 
+ch_listen({address} [, {options}])				*ch_listen()*
```
This seems to imply that there may only be one attached client on this socket. I would have expected there to be a sort of `accept` callback which is passed the `channel` which is created by the client socket event, returning true/false (maybe) as to whether or not to accept the connection.

Perhaps supporting multiple clients on the same port is just something that a text editor shouldn't do, but it doens't seem much of a stretch from 1 to N.

puremourning commented on src/channel.c:
```diff
@@ -724,11 +736,60 @@ channel_connect(
+	server.sin_port = htons(port);
+	if ((host = gethostbyname(channel->ch_hostname)) == NULL)
+	{
+	    ch_error(channel, "in gethostbyname() in channel_open()");
```
does this error include anything that can be used to work out what went wrong?

What happens if we use something like `ch_listen( '0.0.0.0:1234' )` ? How about IPV6 addresses ?

puremourning commented on src/channel.c:
```diff
@@ -724,11 +736,60 @@ channel_connect(
+	return -1;
+    }
+
+    if (STRNCMP(channel->ch_hostname, "unix:", 5) == 0)
```
Is this unrelated to this PR? I mean, is this an undocumented way to use unix domain sockets?

I mean I'm all for supporting unix domain sockets for ch_open() and friends, but is that strictly part of this PR?

puremourning commented on runtime/doc/eval.txt:
```diff
@@ -2448,6 +2448,8 @@ ch_evalraw({handle}, {string} [, {options}])
 ch_getjob({channel})		Job	get the Job of {channel}
 ch_info({handle})		String	info about channel {handle}
+ch_listen({address} [, {options}])
+				Channel	listen to a channel with {address}
```
I think this text could be improved, "Start listening for connections to {address}" perhaps.

puremourning commented on runtime/doc/channel.txt:
```diff
@@ -584,6 +584,17 @@ ch_info({handle})						*ch_info()*
+		"localhost:8765" or ":8765".
+
+		If {options} is given it must be a |Dictionary|.
+		See |channel-open-options|.
```
do all options apply?

puremourning commented on runtime/doc/channel.txt:
```diff
@@ -584,6 +584,17 @@ ch_info({handle})						*ch_info()*
 
+ch_listen({address} [, {options}])				*ch_listen()*
+		Listen to a channel with {address}.  See |channel|.
+		Returns a Channel.  Use |ch_status()| to check for failure.
```
What happens if you write to or read from the channel before the client connects? For that matter, how is the user informed of the client connection?

Or is this API assuming that you'll wait synchronously for a client to connect to this port ? If so I think that's (probably) bad. I find it hard to imagine that all use cases for this would want to block until some client connects. I would have expected it to be more event driven, or at least.... that should be an option? 

For simplicity sake, I can see that a synchronous version might be useful I suppose.

puremourning commented on runtime/doc/channel.txt:
```diff
@@ -584,6 +584,17 @@ ch_info({handle})						*ch_info()*
 			GetChannel()->ch_info()
 
+ch_listen({address} [, {options}])				*ch_listen()*
+		Listen to a channel with {address}.  See |channel|.
```
"Listening to a channel" isn't really what we're doing, we're listening on a port (ok, a socket), there result of which is a channel representing a connection to a client.

puremourning commented on src/channel.c:
```diff
@@ -724,11 +736,60 @@ channel_connect(
+	vim_memset((char *)&server, 0, sizeof(server));
+	server.sin_family = AF_INET;
+	server.sin_port = htons(port);
+	if ((host = gethostbyname(channel->ch_hostname)) == NULL)
```
at least on my system, `man gethostbyname` says 

>      The getaddrinfo(3) and getnameinfo(3) functions are preferred over the gethostbyname(), gethostbyname2(), and gethostbyaddr() functions.

puremourning commented on src/channel.c:
```diff
@@ -3569,6 +3877,48 @@ channel_read(channel_T *channel, ch_part_T part, char *func)
     {
 	if (channel_wait(channel, fd, 0) != CW_READY)
 	    break;
+	if (channel->ch_listen)
```
so is this the only way to accept from the socket? I.e. the user has to do something like:

```viml
let ch = ch_listen( address, #{ callback: function( 'HandleDataOnSocket' ) } )
" see time later...
" this is a blocking call and we'll get data in the callback ? It will return after calling our callback ?
call ch_read( ch )
```

I think we really need an example of how to use this API in practice because it's really not obvious how to use it (as-is) for any of the proposed use cases. Use of `ch_read()` as a proxy for `accept()` when the channel happens to be listening seems surprising to me.

puremourning commented on src/testdir/test_channel.vim:
```diff
@@ -1908,6 +1908,32 @@ func Test_close_lambda()
+  let g:server_received_msg = ch_readraw(a:ch)
+endfunction
+
+func Test_listen()
```
tests seem... optimistic.

the code seems to include a completely new mechanism for using `unix:` to use a domain socket, but no tests for it.

puremourning commented on src/testdir/test_channel.vim:
```diff
@@ -1908,6 +1908,32 @@ func Test_close_lambda()
+  let server = ch_listen('127.0.0.1:' . 12345, {"callback": function("s:test_listen_accept")})
+  if ch_status(server) == "fail"
+    call assert_report("Can't listen channel")
+    return
```
so you set the channel callback to something which calls ch_read, which is what calls `accept` ?

how does it receive subsequent messages ? how to you write to the channel ?

puremourning reviewed (commented):
I was curious about how this would work, so I added some commentary. The API seems a little surprising to me, and it seems to introduce a new unix domain sockets support (which I think is welcome, but undocumented and untested).

skanehira commented:
I want to make a markdown preview plugin using ch_listen.
I think ch_listen brings a  good experience because If we can use ch_listen, we don't need any dependencies to make some plugins.
